File: lab3/java/SearchRandom.java
import java.util.Random;
import java.util.List;
import java.util.LinkedList;

/**
 * Perform a random walk in the graph, hoping to reach the goal.
 * Warning: this class will give up of the random walk
 * reaches a dead end or after 10,000 iterations.
 * So a negative result does not mean there is no path.
 */
public class SearchRandom<V> extends SearchAlgorithm<V> {

    SearchRandom(Graph<V> graph) {
        super(graph);
    }

    @Override
    public SearchResult<V> search(V start, V goal) {
        Random random = new Random();
        int iterations = 0;
        double cost = 0;
        LinkedList<Edge<V>> path = new LinkedList<>();

        V current = start;
        while (iterations < 10_000) {
            iterations++;
            if (current.equals(goal))
                return new SearchResult<>(this.graph, true, start, current, cost, path, iterations);

            List<Edge<V>> neighbours = this.graph.outgoingEdges(current);
            if (neighbours.isEmpty())
                break;

            Edge<V> edge = neighbours.get(random.nextInt(neighbours.size()));
            path.add(edge);
            cost += edge.weight;
            current = edge.end;
        }
        return new SearchResult<>(this.graph, false, start, goal, -1, null, iterations);
    }

}

File: lab3/python/search_random.py
import random
from typing import List

from edge import Edge, V
from search_algorithm import SearchAlgorithm
from search_result import SearchResult


class SearchRandom(SearchAlgorithm[V]):
    """
    Perform a random walk in the graph, hoping to reach the goal.
    Warning: this class will give up if the random walk
    reaches a dead end or after 10,000 iterations.
    So a negative result does not mean there is no path.
    """

    def search(self, start: V, goal: V) -> SearchResult[V]:
        iterations = 0
        cost = 0.0
        path: List[Edge[V]] = []

        current = start
        while iterations < 10_000:
            iterations += 1
            if current == goal:
                return SearchResult(self.graph, True, start, goal, cost, path, iterations)

            neighbours = self.graph.outgoingEdges(current)
            if len(neighbours) == 0:
                break

            edge = random.choice(neighbours)
            path.append(edge)
            cost += edge.weight
            current = edge.end

        return SearchResult(self.graph, False, start, goal, -1, None, iterations)


